### PR TITLE
AWS IAM Managed policy resource

### PR DESCRIPTION
**Pull request by Nirbhay1997**
Signed-off-by: nirbhay1997 <nkumar@progress.com>

### Description

Added Support for Managed Policy/Policies

- TF is not available for the resource

### Check List
Please fill box or appropriate ([x]) or mark N/A.
- [x] New functionality includes integration tests/controls
- [ ] New Terraform resources
- [x] Documentation provided or updated for resources 
- [x] All Integration Tests pass
- [x] All Unit Tests pass
- [x] `rake lint` passes
- [x] All commits have been signed-off for the Developer Certificate of Origin. See <https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco>

### CODE REVIEW

File: libraries/aws_iam_managed_policies.rb [hunk 1]
```diff
@@ -0,0 +1,63 @@
+# frozen_string_literal: true
+
+require 'aws_backend'
+
+class AwsIamManagedPolicies < AwsResourceBase
+  name 'aws_iam_managed_policies'
+  desc 'Verifies settings for a collection AWS Iam Policies'
```
Comment (soumyo13) [inline libraries/aws_iam_managed_policies.rb:7]: Change the description to "Lists all IAM users, groups, and roles that the specified managed policy is attached to."

File: libraries/aws_iam_managed_policies.rb [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    @table = fetch_data(parameters)
+  end
+
+  def fetch_data(parameters = {})
+    iam_policy_rows = []
+    loop do
+      catch_aws_errors do
+        @response = @aws.iam_client.list_entities_for_policy
+      end
+      return [] if !@response || @response.empty?
```
Comment (soumyo13) [inline libraries/aws_iam_managed_policies.rb:40]: return iam_policy_rows

File: libraries/aws_iam_managed_policies.rb [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+                              permissions_boundary_usage_count: p.permissions_boundary_usage_count,
+                              description:                      p.description,
+                              create_date:                      p.create_date,
+                              update_date:                      p.update_date,
+                              tags:                             p.tags,
+
+
+                            }]
+      end
+      break unless @response.is_truncated
```
Comment (soumyo13) [inline libraries/aws_iam_managed_policies.rb:56]: Why this line?

File: libraries/aws_iam_managed_policy.rb [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  def is_attachable?
+    @res.is_attachable
+  end
+
+  def exists?
+    !@arn.nil?
+  end
+
+  def to_s
+    "AWS Iam Policy #{@policy_name}"
```
Comment (soumyo13) [inline libraries/aws_iam_managed_policy.rb:37]: What is @policy_name?

File: libraries/aws_iam_managed_policy.rb [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    raise ArgumentError, "#{@__resource_name__}: policy_arn must be provided" unless opts[:policy_arn] && !opts[:policy_arn].empty?
+    @display_name = opts[:policy_arn]
+    catch_aws_errors do
+      resp = @aws.iam_client.get_policy({ policy_arn: opts[:policy_arn] })
+      @res = resp.policy.to_h
+      create_resource_methods(@res)
+    end
+  end
+
+  def is_attachable?
```
Comment (soumyo13) [inline libraries/aws_iam_managed_policy.rb:28]: Why this function?

File: libraries/aws_iam_managed_policy.rb (hunk 4 above)
Comment (soumyo13) [reply libraries/aws_iam_managed_policy.rb:37]: I think it should be @display_name

And change the string to AWS IAM Policy ARN:

File: test/unit/resources/aws_iam_managed_policy_test.rb [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+  def test_iam_client_non_existing_for_empty_response
+    refute AwsIamManagedPolicy.new(client_args: { stub_responses: true }).exist?
+  end
+end
+
+class AwsIamManagedPolicyHappyPathTest < Minitest::Test
+
+  def setup
+    data = {}
+    data[:method] = :get_profile
```
Comment (soumyo13) [inline test/unit/resources/aws_iam_managed_policy_test.rb:24]: It should be get_policy

File: test/unit/resources/aws_iam_managed_policy_test.rb [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    data[:client] = Aws::IAM::Client
+    @iam_client = AwsIamManagedPolicy.new(client_args: { stub_responses: true }, stub_data: [data])
+  end
+
+  def test_iam_client_exists
+    assert @iam_client.exist?
+  end
+
+  def test_paths
+    assert_equal(@iam_client.paths, 'test1')
```
Comment (soumyo13) [inline test/unit/resources/aws_iam_managed_policy_test.rb:42]: assert_equal(@iam_client.paths, ['test1'])

Comment (soumyo13) [reply test/unit/resources/aws_iam_managed_policy_test.rb:42]: Change the below also.

**Review by soumyo13 — CHANGES_REQUESTED** [review]
Please review the comments.

**Comment by soumyo13** [comment]
Please add the documentation and also the integration tests.

**Comment by Nirbhay1997** [comment]
> Please add the documentation and also the integration tests.

There is no tf resources for managed policy will try to look for alternatives

File: docs/resources/aws_iam_managed_policies.md [hunk 8]
```diff
@@ -0,0 +1,72 @@
+---
+title: About the aws_iam_managed_policies Resource
+platform: aws
+---
+
+# aws\_iam\_policies
```
Comment (soumyo13) [inline docs/resources/aws_iam_managed_policies.md:6]: aws_iam_managed_policies

File: docs/resources/aws_iam_managed_policies.md [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+Use the `aws_iam_managed_policies` InSpec audit resource to test properties of a collection of AWS IAM Policies.
+
+## Syntax
+
+`aws_iam_managed_policies` Resource returns a collection of IAM Policies and allows testing of that collection.
```
Comment (soumyo13) [inline docs/resources/aws_iam_managed_policies.md:12]: `aws_iam_managed_policies` resource returns a collection of IAM Policies and allows testing of that collection.

File: docs/resources/aws_iam_managed_policies.md [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+This resource allows filtering by scope.
+To list only AWS managed policies, set `Scope` to `AWS`. To list only the customer managed policies in your AWS account, set `Scope` to `Local`. If scope is not supplied `ALL` policies are returned.
+
+See also the [AWS documentation on IAM Policy](https://docs.aws.amazon.com/IAM/latest/UserGuide/access_policies.html).
+
+## Properties
+
+|Property              | Description|
+| ---                  | --- |
+|arns                  | The ARN identifier of the specified policy. |
```
Comment (soumyo13) [inline docs/resources/aws_iam_managed_policies.md:29]: Update the descriptions.

Comment (soumyo13) [reply docs/resources/aws_iam_managed_policies.md:29]: https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/iam.html#IAM.Client.get_policy

File: docs/resources/aws_iam_managed_policies.md [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+## Matchers
+
+For a full list of available matchers, please visit our [matchers page](https://www.inspec.io/docs/reference/matchers/).
+
+#### exist
+
+The control will pass if the describe returns at least one result.
+
+Use `should_not` to test the entity should not exist.
```
Comment (soumyo13) [inline docs/resources/aws_iam_managed_policies.md:58]: Change the line.

File: docs/resources/aws_iam_managed_policies.md [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+To list only AWS managed policies, set `Scope` to `AWS`. To list only the customer managed policies in your AWS account, set `Scope` to `Local`. If scope is not supplied `ALL` policies are returned.
+
+See also the [AWS documentation on IAM Policy](https://docs.aws.amazon.com/IAM/latest/UserGuide/access_policies.html).
+
+## Properties
+
+|Property              | Description|
+| ---                  | --- |
+|arns                  | The ARN identifier of the specified policy. |
+|policy\_ids           | The policy ids. |
```
Comment (soumyo13) [inline docs/resources/aws_iam_managed_policies.md:30]: Update the description.

File: libraries/aws_iam_managed_policies.rb [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+    super(opts)
+    validate_parameters
+    @table = fetch_data
+  end
+
+  def fetch_data
+    iam_policy_rows = []
+    loop do
+      catch_aws_errors do
+        @response = @aws.iam_client.list_policies(max_items: 1000)
```
Comment (Nirbhay1997) [inline libraries/aws_iam_managed_policies.rb:39]: TODO

**Review by IanMadd — APPROVED** [review]

**Review by soumyo13 — COMMENTED** [review]
add the terraform.

**Review by soumyo13 — APPROVED** [review]

File: docs/resources/aws_iam_managed_policies.md [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+## Syntax
+
+The `aws_iam_managed_policies` resource returns a collection of IAM managed policies and allows testing of that collection.
+
+    describe aws_iam_managed_policies do
+      its('policy_names') { should include('POLICY_NAME') }
+    end
+
+This resource allows filtering by scope.
+To list only AWS managed policies, set `Scope` to `AWS`. To list only the customer managed policies in your AWS account, set `Scope` to `Local`. If scope is not supplied `ALL` policies are returned.
```
Comment (IanMadd) [inline docs/resources/aws_iam_managed_policies.md:19]:
```suggestion
To list only AWS-managed policies, set `Scope` to `AWS`. To list only the customer-managed policies in your AWS account, set `Scope` to `Local`. If a scope is not supplied, `ALL` policies are returned.
```

Comment (IanMadd) [reply docs/resources/aws_iam_managed_policies.md:19]: Can you add an example of this.

I assume it's something like: `describe aws_iam_managed_policies(scope: 'aws' ) do`

File: docs/resources/aws_iam_managed_policy.md [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+### Test that a specific policy name exists
+
+    describe aws_iam_managed_policy(policy_arn: 'POLICY_ARN') do
+      its('policy_name') { should eq 'POLICY_NAME' }
+    end
+
+### Test that a specific policy ID exists
+
+    describe aws_iam_managed_policy(policy_arn: 'POLICY_ARN') do
+      its('polict_id') { should eq 'POLICY_ID'}
```
Comment (IanMadd) [inline docs/resources/aws_iam_managed_policy.md:51]:
```suggestion
      its('policy_id') { should eq 'POLICY_ID'}
```

File: docs/resources/aws_iam_managed_policies.md (hunk 14 above)
Comment (Nirbhay1997) [reply docs/resources/aws_iam_managed_policies.md]: Thanks, Adding

File: docs/resources/aws_iam_managed_policies.md [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+
+    describe aws_iam_managed_policies do
+      its('policy_names') { should include('POLICY_NAME') }
+    end
+
+This resource allows filtering by scope, which are:
+
+- To list only AWS-managed policies, set `Scope` to `AWS`. 
+- To list only the customer-managed policies in your AWS account, set `Scope` to `Local`.
+- If a scope is not provided, `ALL` policies are returned.
```
[the remark this answers is not in the document]
Comment (soumyo13) [reply docs/resources/aws_iam_managed_policies.md:22]: @Nirbhay1997 Can you check this one?

[the remark this answers is not in the document]
Comment (Nirbhay1997) [reply docs/resources/aws_iam_managed_policies.md:22]: @IanMadd   Yes it(all) can be passed,  but if nothing is passed in scope by default it takes 'all'.

**Review by dkumaras — APPROVED** [review]

**Review by IanMadd — APPROVED** [review]

**Pull request merged**